### PR TITLE
Lean 3 backend (take 2)

### PR DESCRIPTION
**Pull request by jroesch**
This set of commits implements a compiler from Lean's core expression type to C++. This is still very much a prototype and needs more testing and features.

I am closing the previous PR and opening a new one since I changed branches, and this second attempt is more tightly integrated with the Lean 3 redesign. 

I think there is still a little bit of work left before this code should be merged, I'm currently streamlining the CLI and code, and trying to enable passes to be written in Lean itself.  

Previous PR: https://github.com/leanprover/lean/pull/1055

### CODE REVIEW

**Comment by leodemoura** [comment]
I will merge it during the weekend. I'm flying back to Seattle tomorrow.

**Comment by jroesch** [comment]
Going to do a code review for myself.

File: library/init/wf.lean [hunk 1]
```diff
@@ -65,6 +65,10 @@ namespace well_founded
   definition fix [Hwf : well_founded R] (F : Πx, (Πy, R y x → C y) → C x) (x : A) : C x :=
   fix_F F x (apply Hwf x)
 
+  -- Well-founded FO fixpoint
```
Comment (jroesch) [inline library/init/wf.lean]: Remove this.

File: library/system/IO.lean [hunk 2]
```diff
@@ -13,6 +14,14 @@ constant put_str : string → IO unit
 constant put_nat : nat → IO unit
 constant get_line : IO string
 
+constant forever : IO unit -> IO unit
```
Comment (jroesch) [inline library/system/IO.lean:17]: Forever was an experiment, most likely remove.

File: src/kernel/abstract.h [hunk 3]
```diff
@@ -51,6 +51,18 @@ inline expr Pi(expr const & local, expr const & b, bool use_cache = true) { retu
 inline expr Pi(std::initializer_list<expr> const & locals, expr const & b, bool use_cache = true) { return Pi(locals.size(), locals.begin(), b, use_cache); }
 template<typename T> expr Pi(T const & locals, expr const & b, bool use_cache = true) { return Pi(locals.size(), locals.data(), b, use_cache); }
 
+/**
```
Comment (jroesch) [inline src/kernel/abstract.h:54]: Either finish or remove this.

File: src/library/blast/options.cpp [hunk 4]
```diff
@@ -171,6 +171,7 @@ void initialize_options() {
                              "we have this threshold because in the worst case this procedure may take "
                              "an exponetial number of steps");
 }
+
```
Comment (jroesch) [inline src/library/blast/options.cpp]: Remove this.

File: src/library/compiler/anf_transform.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
+/*
+Copyright (c) 2015 Microsoft Corporation. All rights reserved.
```
Comment (jroesch) [inline src/library/compiler/anf_transform.cpp]: Update Copyright.

File: src/library/compiler/anf_transform.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/*
+Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Leonardo de Moura
+*/
+#pragma once
+#include "kernel/environment.h"
+namespace lean {
+/** \brief Annotate locations where control flow of a function exits.
```
Comment (jroesch) [inline src/library/compiler/anf_transform.h]: Update comment.

File: src/library/compiler/annotate_return.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
+/*
+Copyright (c) 2015 Microsoft Corporation. All rights reserved.
```
Comment (jroesch) [inline src/library/compiler/annotate_return.cpp]: Fix copyright.

File: src/library/compiler/annotate_return.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/*
+Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Leonardo de Moura
```
Comment (jroesch) [inline src/library/compiler/annotate_return.h]: Fix copyright.

File: src/library/compiler/cf.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Jared Roesch
+*/
+#pragma once
+#include "kernel/environment.h"
+
+namespace lean {
```
Comment (jroesch) [inline src/library/compiler/cf.h:11]: Update comment.

File: src/library/compiler/cpp_compiler.cpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+#include <fstream>
+#include <iostream>
+#include <iomanip>
+#include <utility>
+#include "cpp_compiler.h"
+#include "kernel/environment.h"
+#include "sys/wait.h"
+
+namespace lean {
+  process::process(std::string n): m_proc_name(n), m_args() {
```
Comment (jroesch) [inline src/library/compiler/cpp_compiler.cpp]: Factor out process into its own utility class.

File: src/library/compiler/native_compiler.cpp [hunk 11]
```diff
@@ -0,0 +1,831 @@
...
+#include <stdlib.h>
+#include <unistd.h>
+
+static std::string* g_lean_install_path;
+
+namespace lean {
+
+// Helper functions for setting up the install path on boot-up.
+// TODO: this is not currently cross platform
+void set_install_path(std::string s) {
```
Comment (jroesch) [inline src/library/compiler/native_compiler.cpp:50]: Clean-up and unify this code with `lean_path.cpp`.

File: src/library/compiler/native_compiler.cpp [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+
+      unsigned arity;
+      this->m_arity_map.for_each([&] (name const & n, unsigned const & i) {
+          // std::cout << "entry: " << n.to_string("+") <<  (n == n_) << std::endl;
+          if (n == n_) {
+            arity = i;
+          }
+      });
+
+      // std::cout << "name:" << n.to_string("&") << std::endl;
```
Comment (jroesch) [inline src/library/compiler/native_compiler.cpp]: Fix this bug.

File: src/library/compiler/native_compiler.cpp [hunk 13]
```diff
@@ -0,0 +1,831 @@
...
+    }
+
+    used_names.m_used_names.for_each([&] (name const & n) {
+        std::cout << n << std::endl;
+        // TODO: unify this
+        if (auto builtin = get_builtin(n)) {
+             // std::cout << "extern fn" << n << std::endl;
+             extern_fns.push_back(builtin.value());
+        } else if (auto i = get_vm_builtin_cases_idx(env, n)) {
+             auto arity = 2; // get_vm_builtin_cases()
```
Comment (jroesch) [inline src/library/compiler/native_compiler.cpp]: This was crashing, when the proper code is placed here. Should fix this.

File: src/library/compiler/used_names.cpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
+/*
```
Comment (jroesch) [inline src/library/compiler/used_names.cpp:1]: Rename to used_defs, I accidentally made a mistake naming this.

File: src/shell/lean.cpp [hunk 15]
```diff
@@ -165,13 +167,16 @@ static struct option g_long_options[] = {
...
+    {"native",       required_argument, 0, 'n'},
     {"export",       required_argument, 0, 'E'},
     {"export-all",   required_argument, 0, 'A'},
     {"memory",       required_argument, 0, 'M'},
     {"trust",        required_argument, 0, 't'},
     {"discard",      no_argument,       0, 'r'},
     {"to_axiom",     no_argument,       0, 'X'},
     {"profile",      no_argument,       0, 'P'},
+    {"compile",      optional_argument, 0, 'C'},
+    // {"main_fn",      required_argument, 0, 'M'}, optional_arg?
```
Comment (jroesch) [inline src/shell/lean.cpp:180]: remove this commented out argument

File: src/shell/lean.cpp [hunk 16]
```diff
@@ -270,10 +280,12 @@ int main(int argc, char ** argv) {
     keep_theorem_mode tmode = keep_theorem_mode::All;
     options opts;
     std::string output;
+    std::string native_output;
     std::string cache_name;
     std::string index_name;
     optional<unsigned> line;
     optional<unsigned> column;
+    optional<std::string> compiler_target;
```
Comment (jroesch) [inline src/shell/lean.cpp]: remove this variable

File: src/shell/lean.cpp [hunk 17]
```diff
@@ -385,6 +401,10 @@ int main(int argc, char ** argv) {
         case 'E':
             export_txt = std::string(optarg);
             break;
+        case 'C':
+            // compiler_target = std::string(optarg);
```
Comment (jroesch) [inline src/shell/lean.cpp]: remove this code

File: src/util/dynamic_loader.cpp [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+*/
+
+#include <string>
+#include "dynamic_loader.h"
+
+namespace lean {
+
+dynamic_library::dynamic_library(std::string library_path):
+m_name(library_path), m_handle(nullptr) {
+    m_handle =  dlopen("file.dylib", RTLD_LAZY | RTLD_GLOBAL);
```
Comment (jroesch) [inline src/util/dynamic_loader.cpp]: Fix this code, and link it up to the loading.

**Comment by leodemoura** [comment]
Github closed this pull request after I deleted the `lean3` branch :(
The master branch is now the old `lean3` branch.
Could you open it again?

**Pull request closed without merging**